lean4.html.markdown

```diff
@@ -0,0 +1,587 @@
+`False`. Once, we have `False`, we can construct anything, and specifically `q`.
+-/
+
+/-
+Same proof, using tactics.
+-/
+theorem contrapositive_reverse : (¬q → ¬p) → (p → q) := by
+  intro hnqnp
+  intro hp
+  have emq := Classical.em q
+  cases emq
+  case inl _ => assumption
+  case inr h => specialize hnqnp h ; contradiction
+
+/-
+To illustrate how we can define an work with axiomatic systems. Here is a
+definition of Groups and some proofs directly translated from "Topics in
+Algebra" by Herstein, Second edition.
+-/
+
+/-
+A class defines a set of axioms
+-/
+class Group (G : Type u) where
+  op : G → G → G
+  assoc : ∀ a b c : G, op (op a b) c = op a (op b c)
+  e : G
+  identity: ∀ a : G, op a e = a ∧ op e a = a
+  inverse: ∀ a : G, ∃ b : G, op a b = e ∧ op b a = e
+
+/-
+Let us introduce some notation to make this convenient.
+-/
+open Group
+infixl:70 " * " => op
+
+/-
+`G` will always stand for a group. Variables `a b c` will be elements of that
+group.
+-/
+variable [Group G] {a b c : G}
+
+def is_identity (e' : G) := ∀ a : G, (a * e' = a ∧ e' * a = a)
+
+/-
+We prove that the identity element is unique.
+-/
+theorem identity_element_unique : ∀ e' : G, is_identity e' → e' = e := by
+  intro e'
+  intro h
+  specialize h e
+  obtain ⟨h1, _⟩ := h
+  have h' := identity e'
+  obtain ⟨_, h2⟩ := h'
+  apply Eq.trans (Eq.symm h2) h1
+/-
+Note that we used the `identity` axiom.
+-/
+
+/-
+Left cancellation. We have to use both `identity` and `inverse` axioms from
+`Group`.
+-/
+theorem left_cancellation : ∀ x y : G, a * x = a * y → x = y := by
+  have h1 := inverse a
+  obtain ⟨ai, a_inv⟩ := h1
+  obtain ⟨_, h2⟩ := a_inv
+  intro x y
+  intro h3
+  calc
+    x = (e : G) * x  := Eq.symm (identity x).right
+    _ = ai * a * x   := by rw [h2]
+    _ = ai * (a * x) := by rw [assoc]
+    _ = ai * (a * y) := by rw [h3]
+    _ = ai * a * y   := by rw [← assoc]
+    _ = (e : G) * y  := by rw [h2]
+    _ = y            := (identity y).right
+
+/-
+Mutually recursive definitions have to be carefully constructed. In particular,
+the hypotheses when recursing should imply well-foundedness.
+
+The game of Nim with two heaps.
+-/
+abbrev between (lower what upper : Nat) : Prop := lower ≤ what ∧ what ≤ upper
+
+mutual
+  def Alice : Nat → Nat → Prop
+    | n1, n2 =>
+      ∃ k, (between 1 k n1 ∧ between 1 k n1 → Bob (n1-k) n2)
+         ∨ (between 1 k n2 ∧ between 1 k n2 → Bob n1 (n2-k))
+
+  def Bob : Nat → Nat → Prop
+    | n1, n2 =>
+      ∀ k, (between 1 k n1 → Alice (n1-k) n2)
+         ∧ (between 1 k n2 → Alice n1 (n2-k))
+end
+
+example : Bob 0 0 := by
+  intro k
+  induction k
+  case zero =>
+    constructor
+    intro ; contradiction
+    intro ; contradiction
+  case succ =>
+    constructor
+    intro a ; have := a.right ; contradiction
+    intro a ; have := a.right ; contradiction
+
+/-
+Programming algorithms in Lean is harder.
+
+This is because we have to convince Lean of termination. Here's a simple
```

Comment:
This is not true. Lean accepts `partial` definitions that can be computed with just fine; just can't be unfolded to reason with.